### PR TITLE
C++: Mishandling Japanese Era and Leap Year in calculations

### PR DESCRIPTION
**Pull request by denislevin**
These queries check for various scenarios where the dates may be mishandled.
* Calculations not taking into account possible leap year
* Japanese dates assuming and hard-coding current (or now previous) era Heisei

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/commons/DateTime.qll [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      exists(Field f, StructLikeClass struct |
+        f.getAnAccess() = this
+        and struct.getAField() = f
+      )
+   }
+}
+
+/**
+ * abstract class of type FieldAccess where access is to a day of the month field of the struct
+ * This is to be derived form for a specific struct's day of the month field access
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/commons/DateTime.qll]: "derived form" -> "derived from" (here and below two more times)

File: cpp/ql/src/Likely Bugs/Leap Year/UncheckedLeapYearAfterYearModification.ql [hunk 2]
```diff
@@ -0,0 +1,61 @@
+/**
+ * @name Year field changed using an arithmetic operation without checking for leap year
+ * @description A that field that represents a year is being modified by an arithmetic operation, but no proper check for leap year can be detected afterwards.
```
Comment (semmledocs-ac) [inline cpp/ql/src/Likely Bugs/Leap Year/UncheckedLeapYearAfterYearModification.ql]: "A that field that" - typo

Comment (semmledocs-ac) [reply cpp/ql/src/Likely Bugs/Leap Year/UncheckedLeapYearAfterYearModification.ql]: "proper check for leap year" 
change to:
"proper check for leap years"

File: cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.qhelp [hunk 3]
```diff
@@ -0,0 +1,36 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+  <include src="LeapYear.qhelp" />
+  
+<p>When using a function that transform a date structure, and the year on the input argument for teh API has been manipulated, it is important to check for the return value of the function to make sure it succeeded.</p>
```
[the remark this answers is not in the document]
Comment (semmledocs-ac) [reply cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.qhelp]: "a function that transform"
change to:
"a function that transforms"

File: cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.qhelp [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  <li><code>TzSpecificLocalTimeToSystemTime</code></li>
+  <li><code>TzSpecificLocalTimeToSystemTimeEx</code></li>
+  <li><code>RtlLocalTimeToSystemTime</code></li>
+  <li><code>RtlTimeToSecondsSince1970</code></li>
+  <li><code>_mkgmtime</code></li>
+</list>
+  
+</overview>
+<recommendation>
+<p>When calling an API that transforms a date variable that was manipulated, always check for the return value to verify if the API call succeeded or failed.</p>
```
Comment (semmledocs-ac) [inline cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.qhelp]: "to verify if the API call succeeded or failed."
change to:
"to determine whether the API call succeeded or failed."
or perhaps:
"to verify that the API call succeeded."

File: cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.ql [hunk 5]
```diff
@@ -0,0 +1,109 @@
+/**
+ * @name Year field changed using an arithmetic operation is used on an unchecked time conversion function
+ * @description A year field changed using an arithmetic operation is used on a time conversion function, but the return value of the function is not check for success or failure
```
Comment (semmledocs-ac) [inline cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.ql]: "is not check for success" > "is not checked for success"

File: cpp/ql/src/Likely Bugs/Leap Year/LeapYear.qhelp [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+<p>The Gregorian Calendar, which has become the internationally accepted civil calendar, the leap year rule is: Every year that is exactly divisible by four is a leap year, except for years that are exactly divisible by 100, but these centurial years are leap years if they are exactly divisible by 400.</p>
+<p>A leap year bug occurs when software (in any language) is written without consideration of leap year logic, or with flawed logic to calculate leap years; which typically results in incorrect results.</p>
+<p>The impact of these bugs may range from almost unnoticeable bugs such as an incorrect date, to severe bugs that affect reliability, availability or even the security of the affected system.</p>
+</overview>
+
+<references>
+<li>U.S. Naval Observatory Website - <a href="https://aa.usno.navy.mil/faq/docs/calendars.php"> Introduction to Calendars</a></li>
+<li>Wikipedia - <a href="https://en.wikipedia.org/wiki/Leap_year_bug"> Leap year bug</a> </li>
+<li>Microsoft Azure blog - <a href="https://azure.microsoft.com/en-us/blog/is-your-code-ready-for-the-leap-year/"> Is your code ready for the leap year?</a> </li>
+</references>
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Leap Year/LeapYear.qhelp]: The `qhelp-preview` job failed with `[ERROR] ql/cpp/ql/src/Likely Bugs/Leap Year/LeapYear.qhelp:16:1: XML document structures must start and end within the same entity.`.  I believe it wants you to add a `</qhelp>` tag here.

File: cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.ql [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+class SafeTimeGatheringFunction extends Function {
+    SafeTimeGatheringFunction() {
+        this.getQualifiedName().matches("GetFileTime")
+        or this.getQualifiedName().matches("GetSystemTime")
+        or this.getQualifiedName().matches("NtQuerySystemTime")
+    }
+}
+
+/**
+ * This list of APIs should check for the return value to detect problems durin gthe conversion
```
Comment (semmledocs-ac) [inline cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.ql]: "durin gthe"

**Comment by rdmarsh2** [comment]
Can you run the autoformatter on the QL files?

File: cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.ql [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    or 
+    exists(MonthFieldAccess  mfa, AssignExpr ae |
+            mfa.getQualifier() = var.getAnAccess()
+            and mfa.isModified()
+            and mfa = fcall.getAPredecessor*()
+            and ae = mfa.getEnclosingElement()
+            and ae.getAnOperand().getValue().toInt() = 1 
+        )
+    )
+select fcall, "Return value of $@ function should be checked to verified for any error because variable $@ is not guaranteed to be safe.", trf, trf.getQualifiedName().toString(), var, var.getName()
```
Comment (semmledocs-ac) [inline cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.ql]: "should be checked to verified for any error"
I'm not sure what this should be.
Maybe:
"should be verified to check for any error"

File: cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYear.qhelp [hunk 9]
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+  <include src="LeapYear.qhelp" />
+
+<p>This query helps to detect when a developer allocates an array or other fixed-length data structure such as <code>std::vector</code> with 365 elements – one for each day of the year.</p>
+<p>Since leap years have 366 days, there will be no allocated element on December 31st at the end of a leap year; which will lead to a buffer overflow on a leap-year.</p>
```
Comment (semmledocs-ac) [inline cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYear.qhelp]: "leap-year"> "leap year"

File: cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYear.qhelp [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+</overview>
+<recommendation>
+<p>When allocating memory for storing elements for each day of the year, ensure that the correct amount of elements are being allocated</p>
```
Comment (semmledocs-ac) [inline cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYear.qhelp]: "correct amount of" > "correct number of"

Comment (semmledocs-ac) [reply cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYear.qhelp]: Probably "are being allocated" is better just as "are allocated"

Comment (semmledocs-ac) [reply cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYear.qhelp]: Add a period at the end of the sentence.

File: cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYear.ql [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    or exists( Variable var |
+        var = element
+        and var.getType() instanceof LeapYearUnsafeDaysOfTheYearArrayType
+    )
+    or exists( ConstructorCall cc |
+        element = cc
+        and cc.getTarget().hasName("vector")
+        and cc.getArgument(0).getValue().toInt() = 365
+    )
+select element, "There is an array or std::vector allocation with a hardcoded set of 365 elements, which may indicate number of days of days in a year without considering leap year scenarios"
```
Comment (semmledocs-ac) [inline cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYear.ql]: "hardcoded" > "hard-coded"

Comment (semmledocs-ac) [reply cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYear.ql]: "number of days" > "the number of days"

File: cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYearGood.c [hunk 12]
```diff
@@ -0,0 +1,4 @@
+bool isLeapYear = year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
+int *items = new int[isLeapYear ? 366 : 365];
+// ...
+delete[] items;
```
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYearGood.c]: There is no silver bullet for this pattern. I am afraid that recommending to always use 366-item arrays may hide the fact that the index offset may still be incorrect. The index offset value must also be set properly for leap years, so devs using this pattern should always be aware of leap years to calculate the offset value correctly. We could add a comment to make this clear.

File: cpp/ql/src/Likely Bugs/Leap Year/LeapYear.qhelp (hunk 6 above)
Comment (raulgarciamsft) [reply cpp/ql/src/Likely Bugs/Leap Year/LeapYear.qhelp]: How can we test the qhelp files to avoid these errors in the future?

Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Leap Year/LeapYear.qhelp]: I personally tend to just rely on the `qhelp-preview` job on the pull request, but I can see that an offline tool might be helpful if you're writing a lot of help.

@semmledocs-ac what do you recommend?

File: cpp/ql/src/Likely Bugs/Leap Year/LeapYear.qll [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+  /**
+   * Holds if the field access is a modification, 
+   * and it involves an arithmetic operation  
+   */
+  predicate isModifiedByArithmeticOperation() {
+    this.isModified()
+    and exists( Operation op |
+      op.getAnOperand() = this
+      and ( op instanceof AssignArithmeticOperation
+        or exists( BinaryArithmeticOperation bao |
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Leap Year/LeapYear.qll]: I'm still not sure what this line is for.  The other part catches `year += n`.  Is this intended to catch `year = year + n`?

**Comment by geoffw0** [comment]
Changes LGTM as well.  All of the changes I asked for have been made as well.  I do have an unanswered question above.

**Comment by geoffw0** [comment]
There were a couple of pull request check failures due to issues in the `qhelp`.  Here is my proposed solution: https://github.com/denislevin/ql/pull/3

**Comment by denislevin** [comment]
Please feel free to merge once it's ready! Thank you!

**Pull request merged**